### ISSUE TITLE
The t function in useTranslation() is not replaced when the namespace changes

### ISSUE THREAD

**Issue by volfpeter** [i18next/react-i18next#1517]
Hi,

First of all, thanks for the development and maintenance of this project!

## 🐛 Bug Report

The `t` function in `useTranslation()` is not replaced (`setT` is not called) when the namespace changes (but everything else is ready and unchanged).

## To Reproduce

If the necessary translation files exist, this basic example should reproduce the bug:

```ts
import React, { useState } from "react";
import { useTranslation } from "react-i18next";

type Message = {
    namespace: string;
    key: string;
};

function MessageBar(props: { message: Message }) {
    const { message } = props;
    const { t } = useTranslation(message.namespace);
    return <div>{t(message.key)}</div>;
}

export function App() {
    const [messages, setMessages] = useState<Message[]>([
        { namespace: "firstNS", key: "firstKey" },
        { namespace: "secondNS", key: "secondKey" },
    ]);

    const { t: firstT } = useTranslation("firstNS");
    const { t: secondT } = useTranslation("secondNS");

    const message = messages.length > 0 ? messages[0] : undefined;

    return (
        <div>
            <div>
                First message: <span>{firstT("firstKey")}</span>
            </div>
            <div>
                Second message: <span>{secondT("secondKey")}</span>
            </div>
            <div>
                <button onClick={() => setMessages(messages.slice(1))}>Next Message</button>
            </div>
            {message !== undefined && <MessageBar message={message} />}
        </div>
    );
}
```

## Expected behavior

The correct message is displayed after clicking the "Next Message" button.

Thanks!

**Comment by volfpeter** [comment]
Thanks for adding the `i18` config and for picking up the bug so quickly.

**Comment by adrai** [comment]
@jamuhl do we need to add the namespaces as a dependency here? https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L89
<img width="336" alt="image" src="https://user-images.githubusercontent.com/1086194/173624273-75836ece-3aca-49fe-b0a3-f97e0633b3ce.png">

**Comment by jamuhl** [comment]
https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L59 as resources are passed in all useTranslation does is check isInitialized...=> no loadNamespace call no reset of `t`

an option might be storing previous ns using something like:

```js
// usePrevious hook
import { useRef, useEffect } from 'react';

export const usePrevious = (value, ignore) => {
  const ref = useRef();
  useEffect(() => {
    ref.current = ignore ? ref.current : value;
  }, [value, ignore]);
  return ref.current;
};
```

else if changed and ready here: https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L63 reset `t`

**Comment by volfpeter** [comment]
Hi,

That was super fast! Thanks for all your time.

Well, it could be that our setup is not typical (we have many smaller translation files, it worked very well so far). We ran into a case/feature request where we need to generate messages in a part of the application where it would be unnecessary, inconvenient/impractical to correctly access the translation function, so we decided on a solution where we generate messages similar to the example and only translate and render them where/when necessary.

Thanks for the fix!
Peter
